iothub/service/src/Registry2/RegistryClient.cs

```diff
@@ -0,0 +1,1177 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net;
+using System.Net.Http;
+using System.Net.Http.Headers;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure;
+using Azure.Core;
+using Microsoft.Azure.Devices.Common.Exceptions;
+using Microsoft.Azure.Devices.Http2;
+
+namespace Microsoft.Azure.Devices
+{
+    /// <summary>
+    /// Client class for all registry related operations such as
+    /// adding/updating/getting/deleting device/module identities
+    /// and getting registry statistics.
+    /// </summary>
+    public class RegistryClient : IDisposable
+    {
+        private string _hostName;
+        private IotHubConnectionProperties _credentialProvider;
+        private HttpTransportSettings2 _settings;
+        private HttpClient _httpClient;
+        private HttpRequestMessageFactory _httpRequestMessageFactory;
+
+        private const string DeviceRequestUriFormat = "/devices/{0}";
+        private const string ModulesRequestUriFormat = "/devices/{0}/modules/{1}";
+        private const string ModulesOnDeviceRequestUriFormat = "/devices/{0}/modules";
+        private const string StatisticsUriFormat = "/statistics/devices";
+        private const string AdminUriFormat = "/$admin/{0}";
+        private const string JobsGetUriFormat = "/jobs/{0}";
+        private const string JobsListUriFormat = "/jobs";
+        private const string JobsCreateUriFormat = "/jobs/create";
+
+        /// <summary>
+        /// Creates an instance of RegistryManager, provided for unit testing purposes only.
+        /// </summary>
+        protected RegistryClient()
+        {
+        }
+
+        /// <summary>
+        /// Create an instance of this class that authenticates service requests
+        /// using an IoT hub connection string.
+        /// </summary>
+        /// <param name="connectionString">The IoT hub connection string.</param>
+        /// <param name="transportSettings">The optional HTTP transport settings.</param>
+        public RegistryClient(string connectionString, HttpTransportSettings2 transportSettings = default)
+        {
+            if (string.IsNullOrEmpty(connectionString))
+            {
+                throw new ArgumentException(nameof(connectionString));
+            }
```

Comment:
Shall we take this moment to clean up our error handling to be more explicit?

```suggestion
            if (connectionString == null)
            {
                throw new ArgumentNullException(nameof(connectionString));
            }
            
            if (string.IsNullOrWhiteSpace(connectionString))
            {
                throw new ArgumentException("Cannot be empty or white space.", nameof(connectionString));
            }
```

That way we distinguish between null and invalid but not null.

Also note that ArgumentException's first parameter is `message`, and the second is `paramName`. But for ArgumentNullException the first parameter is `paramName`. :(

Reply:
Sure, but I'll add a helper to do this since that is a lot of logic to repeat in almost every method